### ISSUE TITLE
iTerm nightly compatibility

### ISSUE THREAD

**Issue by jcrben** [wbond/sublime_terminal#89]
I run iTerm nightly (using `homebrew cask`) but I'd been having issues with Sublime Terminal. I [reported it the iTerm developer George Nachman](https://gitlab.com/gnachman/iterm2/issues/3718#note_1724190) and he cleaned up the iTerm.sh file. His version opens in a new tab of the current terminal by default, which I consider to be the best anyway.

The cleaned up version is available at https://gist.github.com/gnachman/b9ae9cf06c44b5a95edf and it seems reasonable to use it, but I'll hold off for feedback before doing a pull request.

**Comment by twolfson** [comment]
I can't really tell the differences but the script looks fine. Can you open a PR?

**Comment by drewlustro** [comment]
your `iTerm.sh` works, thank you @jcrben

**Comment by gino8080** [comment]
I can confirm the @jcrben cleaned version works!

**Comment by patik** [comment]
@jcrben's version works (thanks!), but there is still one minor issue. If iTerm2 is not running, it opens the app to my default (home) directory. If I press the keyboard shortcut in ST3 a second time it will then open a new tab in iTerm to the project directory. 

Just to be clear, iTerm2 should navigate directly to my project folder the first time I press the shortcut from within ST3. And while I'm already nitpicking :blush: the iTerm2 window is launching in the background but it should come right to the forefront. 

I'm running the [version 3 beta](https://iterm2.com/version3.html?src=4). To be fair to this project, SourceTree (which also has an "open terminal" button) also opens the iTerm beta in the background.

**Comment by patik** [comment]
Installing shell integrations with iTerm2 seems to have alleviated most of the issues in my previous comment. It's still not quite perfect; if you had no open tabs when you last quit iTerm2 it will remaining in the background after invoking it from ST3. But if you had an open tab the shortcut in ST3 will bring iTerm2 to the forefront and navigate to the project's directory.

**Comment by twolfson** [comment]
Can anyone verify that this patch works for both the pre-v3 version and post-v3 version?

**Comment by fdeneux** [comment]
@jcrben's iTerm.sh works with shell integration for iTerm2 Build 2.9.20160206.

**Comment by twolfson** [comment]
Alright, according to the beta release website that is the post-v3 version.

https://iterm2.com/version3.html

Can anyone verify this works with the pre-v3 version?

**Comment by maicong** [comment]
It's awesome!

**Comment by matthias-vogt** [comment]
@jcrben cleaned version didn't work for me until I changed

	tell application "System Events"
	    count(processes whose name is "iTerm")
	end tell
…to

	tell application "System Events"
	    count(processes whose name is "iTerm2")
	end tell

using iTerm 2.9.20160313

**Comment by zertosh** [comment]
FWIW, I'm using this script to open a new tab with iTerm2 2.9.20160313:

```applescript
#!/usr/bin/osascript

do shell script "pwd"
set targetFolder to quoted form of POSIX path of result

tell application "/Applications/iTerm.app"
  -- A count of 0 means that iTerm is open with no windows, or that iTerm is
  -- closed. If it's closed, counting the windows opens it.
  if (count of windows) is 0 then
    -- If iTerm was closed, but is now open, it'll have one window.
    -- This prevents opening two tabs at startup.
    if (count of windows) is 1 then
      set newTab to (current tab of first window)
    else
      set newTab to (current tab of (create window with default profile))
    end if
  else
    set newTab to ((create tab with default profile) of first window)
  end if
  tell current session of newTab to write text "cd " & targetFolder
  activate
end tell
```

**Comment by mbigras** [comment]
I just cloned @jcrben's [sublime_terminal](https://github.com/jcrben/sublime_terminal) but it still isn't working for me. I'm  using iTerm Build 2.9.20160510 with Shell Integration. I've also tried @matthias-vogt's suggestion which also didn't work (just to clarify @matthias-vogt, you were recomending to Replace All iTerm -> iTerm2 correct? That's what I tried and it didn't work). I also tried using the up-to-date repo here and tried the same thing: with and without iTerm -> iTerm2 substitution. I've also tried just manually copying and pasting the script from the gist but this hasn't worked either. I'd love some suggestions :)

**Comment by yumyo** [comment]
@jcrben version solved the issue with iTerm2 Build 3.0.0 (OS X 10.11.5 / Sublime 3114)

**Comment by KyleKing** [comment]
Ditto, the iTerm update to 3.0.0 broke the terminal shortcut. @jcrben 's new [iTerm.sh](https://gist.github.com/gnachman/b9ae9cf06c44b5a95edf) file fixed it!

The only change I made was to focus iTerm by adding `activate` into the tell blocks

**Comment by mbigras** [comment]
Also works for me :) @KyleKing thanks for confirming, would you mind posting a link to your modified iTerm.sh with the correctly added `activate`?

**Comment by KyleKing** [comment]
Nothing special, I just tossed in an `activate` call in three places (Too bad GFM doesn't support highlighting a specific line)

```shell
#!/bin/bash

# Modified following this issue: https://github.com/wbond/sublime_terminal/issues/89

CD_CMD="cd "\\\"$(pwd)\\\"" && clear"
if echo "$SHELL" | grep -E "/fish$" &> /dev/null; then
  CD_CMD="cd "\\\"$(pwd)\\\""; and clear"
fi
VERSION=$(sw_vers -productVersion)
OPEN_IN_TAB=0

while [ "$1" != "" ]; do
	PARAM="$1"
	VALUE="$2"
	case "$PARAM" in
		--open-in-tab)
			OPEN_IN_TAB=1
			;;
	esac
	shift
done

if (( $(expr $VERSION '<' 10.7) )); then
	RUNNING=$(osascript<<END
	tell application "System Events"
	    count(processes whose name is "iTerm")
	end tell
END
)
else
	RUNNING=1
fi

if (( ! $RUNNING )); then
	osascript<<END
	tell application "iTerm"
            tell current window
                tell current session of (create tab with default profile)
                    write text "$CD_CMD"
                end tell
            end tell
            activate
	end tell
END
else
	if (( $OPEN_IN_TAB )); then
		osascript &>/dev/null <<EOF
		tell application "iTerm"
                    if (count of windows) = 0 then
                        set theWindow to (create window with default profile)
                        set theSession to current session of theWindow
                    else
                        set theWindow to current window
                        tell current window
                            set theTab to create tab with default profile
                            set theSession to current session of theTab
                        end tell
                    end if
                    tell theSession
                        write text "$CD_CMD"
                    end tell
                    activate
		end tell
EOF
	else
		osascript &>/dev/null <<EOF
		tell application "iTerm"
                    tell (create window with default profile)
                        tell the current session
                            write text "$CD_CMD"
                        end tell
                    end tell
                    activate
		end tell
EOF
	fi
fi
```

**Comment by excetara2** [comment]
Yeah I just swapped yours in and working great as well. Thanks @KyleKing

**Comment by nalinbhardwaj** [comment]
@excetara2 I was having the same issue.
I dont know really know much of shell scripting but I tweaked the script a little bit and it worked. [Check it out](https://gist.github.com/nalinbhardwaj/7759f056f486b322b5f5aa0980b8b6b9).

**Comment by excetara2** [comment]
@nalinbhardwaj  Still doesn't work for me with your updated script but I think the script just opens whatever folder is passed to it so maybe something is wrong with it grabbing my project folder.

**Comment by excetara2** [comment]
@patik I also had similar problems but only noticed when in full screen mode that it switches to another space and puts Iterm in the background. This doesn't happen if Iterm is not in full screen mode but once it has been activated like this it seems to stay and always open in the background after that.

**Comment by horatiorosa** [comment]
I apologize in advance for this basic question, but how exactly do I update the iTerm.sh to the new fix for build 3? I would appreciate any feedback on this.

**Comment by nalinbhardwaj** [comment]
@horatiorosa Just go to the preferences menu in Sublime text, and click "Browse Packages", from there, open the folder called "Terminal". The file iTerm.sh is located there, just replace it's text, and restart Sublime. See below for the Mac screenshot.

![screen shot 2016-06-12 at 11 15 10 pm](https://cloud.githubusercontent.com/assets/6984346/15992837/fe9275b0-30f3-11e6-9a5c-2a749bf25193.png)

**Comment by horatiorosa** [comment]
Thanks for the help @nalinbhardwaj!

**Comment by uneart** [comment]
For me the new iTerm.sh doesn't seem to work. I tried it yesterday and was able to open iTerm but after a restart it won't work anymore for some reason. I tried both versions of @KyleKing and @jcrben but neither of them works.
When clicking "Open Terminal here...", nothing at all happens

I'm using Sublime Text 3, Build 3114 and iTerm2 Build 3.0.2 on OS X 10.11.5. 

Any suggestions what to try next?

**Comment by KyleKing** [comment]
@uneart can you try opening a different folder or file? Is anything logging to the sublime console?

**Comment by uneart** [comment]
@KyleKing I tried various folders, still no success. There is nothing logged in the console.

**Comment by deweller** [comment]
I'm having a similar problem.  Looks like iTerm.sh is failing for me.

```
$ pwd
/Users/dweller/Library/Application Support/Sublime Text 3/Packages/Terminal
$ ./iTerm.sh
41:49: syntax error: Expected end of line but found identifier. (-2741)
```

iTerm Build 3.0.2
Sublime 3114
Terminal 1.12.0

**Comment by inderpreet99** [comment]
It's probably a better idea to place the new iTerm.sh file in `Sublime Text 3/Packages/User` than putting it `Sublime Text 3/Packages/Terminal`. The User dir won't be overwritten in Terminal package updates. Remember to change the `terminal` key/value in Terminal package's User settings to point to the new location of iTerm.sh. Note: it will require an absolute path.

PS: Why can't the developers of this plugin just add this file to the default package? It seems like almost every person using this package and newer iTerm is having to go through hoops to do so.

**Comment by stiweb** [comment]
Uninstall iTerm Build 3.0.2 and install the version 2.1.4 you'll see that everything goes back to work also with ST3. iTerm 3.0.2 is a test releases update many times a year and are occasionally unstable.

**Comment by uneart** [comment]
Finally! The script seems to work after I replaced all occurrences of `iTerm` with `iTerm2`. For some reason after updating iTerm to Build 3.0.2 it is named iTerm2 everywhere now.

@inderpreet99 Good advice! I put the `iTerm.sh` file into `Sublime Text 3/Packages/User/Terminal` and changed the user settings of the Terminal plugin as follows:

```json
{
    "terminal": "Packages/User/Terminal/iTerm.sh",
    "parameters": ["--open-in-tab"]
}
```

I double-checked the path but I still get the error `Terminal: The terminal Packages/User/Terminal/iTerm.sh was not found`. I checked the user settings of other plugins and they reference the path to `Packages` the same way. What am I missing?

**Comment by inderpreet99** [comment]
@uneart The absolute path would be `/Users/username/Library/Application Support/Sublime Text 3/Packages/User/Terminal/iTerm.sh` (check if that's where you put it). I'm not sure if you could get it to work with relative paths like you have.

@stiweb I do see your point about supporting a non-stable iTerm. I actually meant maybe put in `iTerm2-v3.sh`, rather than replacing the standard `iTerm.sh` (works for version 2). Users already have to setup for `iTerm.sh` by putting it in the settings. This would just make them do it for their specific version of iTerm.

**Comment by certainlyakey** [comment]
My two cents: for some reason adding the settings to OS-specific settings file didn't work, adding them to global User settings did. Also absolute or ST-relative paths (`Packages/...`) didn't work, `iTerm.sh` did. I used the @KyleKing version with `iTerm` string replaced to `iTerm2`.

**Comment by elfenars** [comment]
I opened #124 it works for me with iTerm2 build 3.0.4

**Comment by garyking** [comment]
Anyone has code that works with iTerm 3.0.7?

**Comment by garyking** [comment]
Anyone got a script that will also work if iTerm is not already running?

**Comment by twolfson** [comment]
This issue has been resolved by finally landing #98. Please see the new documentation to get started, sorry for the long delay:

https://github.com/wbond/sublime_terminal/tree/1.13.0#iterm2-v3-on-os-x
